File: src/api/setTemplates.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  SetTemplate,
  SetTemplateSummary,
  SetTemplateCard,
  SetTemplateDetail,
  CardType,
  PlaceholderTag,
} from "../types/setTemplate";

export const setTemplatesApi = {
  list: () => invoke<SetTemplateSummary[]>("list_set_templates"),

  get: (id: string) => invoke<SetTemplateDetail>("get_set_template", { id }),

  create: (name: string, notes: string | null) =>
    invoke<SetTemplate>("create_set_template", { name, notes }),

  update: (id: string, name: string, notes: string | null) =>
    invoke<SetTemplate>("update_set_template", { id, name, notes }),

  delete: (id: string) => invoke<void>("delete_set_template", { id }),

  clone: (id: string) => invoke<SetTemplate>("clone_set_template", { id }),

  addCard: (params: {
    setId: string;
    cardType: CardType;
    exerciseId?: string | null;
    placeholderTag?: PlaceholderTag | null;
    placeholderLabel?: string | null;
    durationHintSec?: number | null;
    notes?: string | null;
  }) =>
    invoke<SetTemplateCard>("add_card", {
      setId: params.setId,
      cardType: params.cardType,
      exerciseId: params.exerciseId ?? null,
      placeholderTag: params.placeholderTag ?? null,
      placeholderLabel: params.placeholderLabel ?? null,
      durationHintSec: params.durationHintSec ?? null,
      notes: params.notes ?? null,
    }),

  updateCard: (params: {
    cardId: string;
    exerciseId?: string | null;
    placeholderTag?: PlaceholderTag | null;
    placeholderLabel?: string | null;
    durationHintSec?: number | null;
    notes?: string | null;
  }) =>
    invoke<SetTemplateCard>("update_card", {
      cardId: params.cardId,
      exerciseId: params.exerciseId ?? null,
      placeholderTag: params.placeholderTag ?? null,
      placeholderLabel: params.placeholderLabel ?? null,
      durationHintSec: params.durationHintSec ?? null,
      notes: params.notes ?? null,
    }),

  removeCard: (cardId: string) => invoke<void>("remove_card", { cardId }),

  reorderCards: (setId: string, orderedIds: string[]) =>
    invoke<void>("reorder_cards", { setId, orderedIds }),
};

File: src-tauri/src/commands/set_templates.rs
use sqlx::SqlitePool;
use tauri::State;
use crate::{
    domain::{
        set_template,
        types::{SetTemplateRow, SetTemplateSummaryRow, SetTemplateCardRow, SetTemplateDetail},
    },
    error::AppError,
};

#[tauri::command]
pub async fn list_set_templates(
    pool: State<'_, SqlitePool>,
) -> Result<Vec<SetTemplateSummaryRow>, AppError> {
    set_template::list(&pool).await
}

#[tauri::command]
pub async fn get_set_template(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<SetTemplateDetail, AppError> {
    set_template::get(&pool, &id).await
}

#[tauri::command]
pub async fn create_set_template(
    pool: State<'_, SqlitePool>,
    name: String,
    notes: Option<String>,
) -> Result<SetTemplateRow, AppError> {
    set_template::create(&pool, &name, notes.as_deref()).await
}

#[tauri::command]
pub async fn update_set_template(
    pool: State<'_, SqlitePool>,
    id: String,
    name: String,
    notes: Option<String>,
) -> Result<SetTemplateRow, AppError> {
    set_template::update(&pool, &id, &name, notes.as_deref()).await
}

#[tauri::command]
pub async fn delete_set_template(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<(), AppError> {
    set_template::delete(&pool, &id).await
}

#[tauri::command]
pub async fn clone_set_template(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<SetTemplateRow, AppError> {
    set_template::clone_set(&pool, &id).await
}

#[tauri::command]
pub async fn add_card(
    pool: State<'_, SqlitePool>,
    set_id: String,
    card_type: String,
    exercise_id: Option<String>,
    placeholder_tag: Option<String>,
    placeholder_label: Option<String>,
    duration_hint_sec: Option<i64>,
    notes: Option<String>,
) -> Result<SetTemplateCardRow, AppError> {
    set_template::add_card(
        &pool,
        &set_id,
        &card_type,
        exercise_id.as_deref(),
        placeholder_tag.as_deref(),
        placeholder_label.as_deref(),
        duration_hint_sec,
        notes.as_deref(),
    )
    .await
}

#[tauri::command]
pub async fn update_card(
    pool: State<'_, SqlitePool>,
    card_id: String,
    exercise_id: Option<String>,
    placeholder_tag: Option<String>,
    placeholder_label: Option<String>,
    duration_hint_sec: Option<i64>,
    notes: Option<String>,
) -> Result<SetTemplateCardRow, AppError> {
    set_template::update_card(
        &pool,
        &card_id,
        exercise_id.as_deref(),
        placeholder_tag.as_deref(),
        placeholder_label.as_deref(),
        duration_hint_sec,
        notes.as_deref(),
    )
    .await
}

#[tauri::command]
pub async fn remove_card(pool: State<'_, SqlitePool>, card_id: String) -> Result<(), AppError> {
    set_template::remove_card(&pool, &card_id).await
}

#[tauri::command]
pub async fn reorder_cards(
    pool: State<'_, SqlitePool>,
    set_id: String,
    ordered_ids: Vec<String>,
) -> Result<(), AppError> {
    set_template::reorder_cards(&pool, &set_id, ordered_ids).await
}
